File: CoSim/backend/src/co_sim/schemas/git.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class GitStatusEntry(BaseModel):
    path: str
    staged: str = Field(min_length=1, max_length=1)
    unstaged: str = Field(min_length=1, max_length=1)


class GitStatusResponse(BaseModel):
    entries: list[GitStatusEntry]


class GitAddRequest(BaseModel):
    paths: Optional[list[str]] = None


class GitCommitRequest(BaseModel):
    message: str = Field(min_length=1)


class GitDiffResponse(BaseModel):
    diff: str

File: CoSim/frontend/src/api/git.ts
import { authorizedClient } from './client';

export interface GitStatusEntry {
  path: string;
  staged: string;
  unstaged: string;
}

export interface GitStatusResponse {
  entries: GitStatusEntry[];
}

export interface GitDiffResponse {
  diff: string;
}

export const getGitStatus = async (token: string, workspaceId: string): Promise<GitStatusResponse> => {
  const { data } = await authorizedClient(token).get<GitStatusResponse>(
    `/v1/workspaces/${workspaceId}/git/status`
  );
  return data;
};

export const gitAdd = async (token: string, workspaceId: string, paths?: string[]): Promise<void> => {
  await authorizedClient(token).post(`/v1/workspaces/${workspaceId}/git/add`, {
    paths
  });
};

export const gitCommit = async (token: string, workspaceId: string, message: string): Promise<{ status: string; output: string }> => {
  const { data } = await authorizedClient(token).post<{ status: string; output: string }>(
    `/v1/workspaces/${workspaceId}/git/commit`,
    { message }
  );
  return data;
};

export const gitDiff = async (
  token: string,
  workspaceId: string,
  options: { staged?: boolean; path?: string } = {}
): Promise<GitDiffResponse> => {
  const { data } = await authorizedClient(token).get<GitDiffResponse>(
    `/v1/workspaces/${workspaceId}/git/diff`,
    { params: options }
  );
  return data;
};
